modules/apps/commerce/commerce-checkout-web/src/main/resources/META-INF/resources/js/AddressSubtypeAutocomplete.tsx

```diff
@@ -0,0 +1,399 @@
+					</Autocomplete.Item>
+				)}
+			</Autocomplete>
+
+			{hasError ? (
+				<ClayForm.FeedbackItem>{errorMessage}</ClayForm.FeedbackItem>
+			) : null}
+		</ClayForm.Group>
+	);
+}
+
+function AddressSubtypeAutocomplete({
+	addressSubtypeConfiguration = {
+		billing: '',
+		billingAndShipping: '',
+		shipping: '',
+	},
+	initialAddressId = '0',
+	initialAddressType,
+	initialValue = '',
+	namespace,
+}: IAddressSubtypeAutocompleteProps) {
+	const {
+		observer,
+		onOpenChange: onErrorModalOpenChange,
+		open: openErrorModal,
+	} = useModal({
+		onClose: () => {
+			const useAsBillingCheckbox = document.getElementById(
+				`${namespace}use-as-billing`
+			) as HTMLInputElement;
+
+			if (useAsBillingCheckbox) {
+				useAsBillingCheckbox.checked = false;
+			}
+		},
+	});
+
+	const [addressId, setAddressId] = useState(Number(initialAddressId));
+	const [addressType, setAddressType] = useState(initialAddressType);
+	const [modalCurrentValue, setModalCurrentValue] = useState('');
+	const [currentValue, setCurrentValue] = useState(initialValue);
+	const [hasError, setHasError] = useState(false);
+	const [externalReferenceCode, setExternalReferenceCode] =
+		useState<string>();
+
+	useEffect(() => {
+		if (
+			['billing', '1'].includes(String(addressType || initialAddressType))
+		) {
+			setExternalReferenceCode(addressSubtypeConfiguration.billing);
+
+			return;
+		}
+		else if (
+			['shipping', '3'].includes(
+				String(addressType || initialAddressType)
+			)
+		) {
+			setExternalReferenceCode(addressSubtypeConfiguration.shipping);
+
+			return;
+		}
+
+		setExternalReferenceCode(
+			addressSubtypeConfiguration.billingAndShipping
+		);
+	}, [
+		addressSubtypeConfiguration.billing,
+		addressSubtypeConfiguration.billingAndShipping,
+		addressSubtypeConfiguration.shipping,
+		addressType,
+		initialAddressType,
+	]);
+
+	useEffect(() => {
+		const commerceAddressSelect = document.getElementById(
+			`${namespace}commerceAddress`
+		) as HTMLSelectElement;
+		const useAsBillingCheckbox = document.getElementById(
+			`${namespace}use-as-billing`
+		) as HTMLInputElement;
+
+		const onCommerceAddressChange = (event: Event) => {
+			setAddressId((event.target as any).value);
+
+			const option = Array.from(
+				(event.target as any).children as HTMLOptionsCollection
+			).filter((item) => item.value === (event.target as any).value)[0];
+
+			setAddressType(option.dataset.type as string);
+			setCurrentValue(option.dataset.subtype || '');
+			setHasError(false);
+		};
+
+		const onUseAsBillingCheckboxChange = (event: Event) => {
+			if (!Number(addressId)) {
+				setExternalReferenceCode(
+					(event.target as HTMLInputElement).checked
+						? addressSubtypeConfiguration.billingAndShipping
+						: addressSubtypeConfiguration.shipping
+				);
+
+				if (currentValue) {
+					setCurrentValue('');
+					setHasError(true);
+				}
+
+				return;
+			}
+
+			if (
+				Number(addressId) &&
+				['shipping', '3'].includes(String(addressType)) &&
```

Comment:
I'm more and more convinced a map of constants could make the code way cleaner here. [NB]

Reply:
replaced with constants